File: My Answer/E4.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class E4 {
    private static final Scanner scanner = new Scanner(System.in);

    public static void Q1(){
        int[] vendas = {1500, 2000, 809, 3500, 1200};
        int total_vendas = 0;
        int quantidade_dias = vendas.length;
        int maior_venda = 0;
        int menor_venda = 100000;
        for(int venda : vendas){
            total_vendas += venda;
            if(maior_venda < venda) maior_venda = venda;
            if(menor_venda > venda) menor_venda = venda;
        }
        double media_vendas = (double)total_vendas/(double)quantidade_dias;
        System.out.println(
            "Relatório:\n - Total de vendas: " + total_vendas
                + "\n - Média de Vendas: " + Math.round(media_vendas)
                + "\n - Maior Venda: " + maior_venda
                + " \n - Menor Venda: " + menor_venda
            );
    }

    public static void Q2(String novoItem, String encontrarItem, String itemAtualizar, String itemAtualizado){
        List<String> estoque = new ArrayList<>(Arrays.asList("monitor", "teclado", "mouse", "headset"));
        System.out.println("Inserir item");
        estoque.add(novoItem);
        System.out.println("Estoque atual: " + estoque);
        int c = 0;
        System.out.println("Encontrar item");
        boolean flag = false;
        for(String item : estoque){
            if(item.equals(encontrarItem)){
                System.out.println(item + " encontrado na posição "+ c + ".");
                flag = true;
                break;
            }
            c++;
        }
        System.out.println((flag) ? "Item encontrado.":"Item não encontrado.");
        System.out.println("Atualizar Item");
        for(int i = 0; i < estoque.size(); i++) {
            if(estoque.get(i).equals(itemAtualizar)) {
                estoque.set(i, itemAtualizado); // Atualiza a posição real na lista
                System.out.println("Item Atualizado");
            }
        }
        System.out.println("Estoque atual: " + estoque);
    }

    public static void Q3(){
        List<Integer> frete = new ArrayList<>(Arrays.asList(50, 80, 20, 150, 40));
        //frete.sort(null);
        frete.sort(Collections.reverseOrder());
        System.out.println(frete);
        List<Integer> top_fretes = frete.subList(0, 2);
        System.out.println(top_fretes);
    }

    public static void Q4(String[] novas_cidades, String encontrar_cidades){
        List<String> cidades = new ArrayList<>(Arrays.asList("Sao Paulo", "Campinas", "Jundiai", "Sorocaba"));
        System.out.println("Cidades registradas: " + cidades);
        System.out.println("Adicionar novas cidades: " + Arrays.asList(novas_cidades));
        for(String cidade : novas_cidades){
            cidades.add(cidade);
        }
        System.out.println("Cidades registradas: " + cidades);
        System.out.println("Encontrar cidade: " + encontrar_cidades);
        boolean flag = false;
        int c = 0;
        for(int i = 0; i < cidades.size(); i++){
            String cidade = cidades.get(i);
            //System.out.println( cidade + " " + (cidade == encontrar_cidades) + " " + encontrar_cidades);
            if(cidade.equals(encontrar_cidades)){
                System.out.println(cidade + " è a "+ c + "ª cidade na rota.");
                flag = true;
                break;
            }
            c++;
        }
        if(!flag) System.out.println(encontrar_cidades + " não foi encontrada.");
    }

    private static String removeAllButLastDot(String s) {
        int last = s.lastIndexOf('.');
        if (last == -1) return s.replace(".", "");
        String before = s.substring(0, last).replace(".", "");
        String after = s.substring(last);
        if((s.length() - last) > 3) after = after.replace(".", "") + ".00";
        return before + after;
    }

    public static double extractDoubleValue(){
        while(true){
            System.out.print("Digite o novo preço do produto: ");
            String newPrice = scanner.nextLine();
            newPrice = newPrice.toUpperCase()
                                       .replace("RS","")
                                       .replace("R$","")
                                       .trim();
            double numPrice = 0.0;
            try {
                numPrice = Double.parseDouble(newPrice);
                return numPrice;
                
            } catch (NumberFormatException e) {
                newPrice = newPrice.replace(",", ".");
                newPrice = removeAllButLastDot(newPrice);
                try {
                    numPrice = Double.parseDouble(newPrice);
                    return numPrice;
                } catch (NumberFormatException e3) {
                    System.out.println("❌ Formato inválido! Digite um valor numérico válido.");
                }
            }
        }
    }

    public static void Q5(){
        List<Double> precos = new ArrayList<>(Arrays.asList(100.0, 200.0, 300.0));
        List<String> vinhos = new ArrayList<>(Arrays.asList("Branco", "Tinto", "Champagne"));
        System.out.println("Vinhos: " + vinhos + "\nPreços: " + precos);
        String produtoAtualizar = "";
        double novoPreco = 0.0;
        while(true){
            System.out.print("Digite o nome do produto: ");
            produtoAtualizar = scanner.nextLine();
            int numindex = -1;
            for(int i = 0; i < vinhos.size(); i++){
                String produto = vinhos.get(i);
                System.out.println(produtoAtualizar + produto + (produto.equals(produtoAtualizar)));
                if(produto.equals(produtoAtualizar)){
                    System.out.println("Produto Encontrado");
                    numindex = i;
                    break;
                }
            }
            if(numindex != -1){
                novoPreco = extractDoubleValue();
                precos.set(numindex, novoPreco);
                break;
            }else{
                System.out.println("Produto Inválido ou Inexistente.");
            }

        }
        
        System.out.println("Vinhos: " + vinhos + "\nPreços: " + precos);

    }

    public static void main(String[] args){
        System.out.println("Dashboard de Vendas (Análise de Dados)");
        Q1();
        System.out.println("Gestão de Estoque (Edição e Verificação)");
        Q2("webcam", "impressora", "teclado", "teclado mecanico");
        System.out.println("Organização de Preços (Ordenação e Slicing)");
        Q3();
        System.out.println("Sistema de Logística (Busca e Extensão)");
        String[] novas_cidades = {"Itu", "Valinhos"};
        Q4(novas_cidades, "Sorocaba");
        System.out.println("Atualização de Preços Interativa (Input + Lista)");
        Q5();
        
    }
    
}

File: My Answer/E4.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <iomanip>
#include <algorithm>
#include <cctype>
#include <locale>
#include <regex>
#include <cmath>
#include <stdexcept>

using namespace std;

void replaceAll(string& str, const string& from, const string& to) {
    if(from.empty()) return;
    size_t start_pos = 0;
    while((start_pos = str.find(from, start_pos)) != string::npos) {
        str.replace(start_pos, from.length(), to);
        start_pos += to.length(); // Avança para não pegar a mesma string substituída
    }
}

void trimString(string &s) {
    s.erase(s.begin(), find_if(s.begin(), s.end(), [](unsigned char ch) { return !isspace(ch); }));
    s.erase(find_if(s.rbegin(), s.rend(), [](unsigned char ch) { return !isspace(ch); }).base(), s.end());
}

string removeAllButLastDot(string s){
    int last = s.find_last_of('.');
    if(last == string::npos) {
        s.erase(remove(s.begin(), s.end(), '.'), s.end()); // Remove todos os pontos
        return s;
    }
    
    string before = s.substr(0, last);
    before.erase(remove(before.begin(), before.end(), '.'), before.end());
    
    string after = s.substr(last);
    if((s.length() - last) > 3) {
        after.erase(remove(after.begin(), after.end(), '.'), after.end());
        after += ".00";
    }
    return before + after;
}

double extractBillingValue(){
    string billing;
    getline(cin >> ws, billing);
    transform(billing.begin(), billing.end(), billing.begin(), ::toupper);
    replaceAll(billing, "RS", "");
    replaceAll(billing, "R$", "");
    trimString(billing);
    replaceAll(billing, ",", ".");
    return stod(removeAllButLastDot(billing));
}

void Q1(){
    int vendas[] = {1500, 2000, 800, 3500, 1200};
    int sum = 0;
    float avr = 0;
    int maior_venda = 0, menor_venda = 100000, n = size(vendas);
    #pragma omp parallel for reduction(+:sum, avr) reduction(max:maior_venda) reduction(min:menor_venda)
    for (int i = 0; i < n; i++){
        sum += vendas[i];
        avr += (float)vendas[i]/n;
        if(maior_venda < vendas[i]) maior_venda = vendas[i];
        if(menor_venda > vendas[i]) menor_venda = vendas[i];
    }
    cout << "Relatório:\n - Total de vendas: " << sum
                << "\n - Média de Vendas: " << avr << setprecision(2)
                << "\n - Maior Venda: " << maior_venda
                << " \n - Menor Venda: " << menor_venda << endl;
}

void Q2(){
    vector<string> estoque = {"monitor", "teclado", "mouse", "headseat"};
    string novoItem = "webcam";
    string nomeAtualizar = "teclado";
    string atualizarNome = "teclado_maníaco";
    string removerItem = "mouse";
    for (const string& item : estoque) cout << "| " << item << " |";
    cout<<endl;
    estoque.push_back(novoItem);
    for (const string& item : estoque) cout << "| " << item << " |";
    cout<<endl;
    for (int i = 0; i < estoque.size(); i++){
        if(estoque[i] == nomeAtualizar){
            estoque[i] = atualizarNome;
            cout<<"Item atualizado!\n";
            break;
        }
    }
    for (const string& item : estoque) cout << "| " << item << " |";
    cout<<endl;
    for (int i = 0; i < estoque.size(); i++){
        if(estoque[i] == removerItem){
            estoque.erase(estoque.begin() + i);
            cout<<"Item removido!\n";
            break;
        }
    }
    for (const string& item : estoque) cout << "| " << item << " |";
    cout<<endl;
    
}

void Q3(){//Ordenação de Preços
    vector<int> precos = {50, 80, 20, 150, 40};
    int min = 100000, max = 0;
    int i = 0, j = precos.size()-1;
    for(const int& preco : precos) cout << preco << " ";
    cout<<endl;
    while(i <= j){
        int temp = 0;
        if(precos[i] > precos[j]){
            temp = precos[j];
            precos[j] = precos[i];
            precos[i] = temp;
            j--;
        }else{
            j--;
        }
        cout << i << " " << j << endl;
        if(i == j) {
            j = precos.size() -1;
            i++;
            if(i == precos.size() - 1) break;
        }
        for(const int& preco : precos) cout << preco << " ";
        cout<<endl;
    }
    cout<<"Lista Ordenada:\n";
    for(const int& preco : precos) cout << preco << " ";
    cout<<endl;
    vector<int> top_fretes(precos.end()-2, precos.end());
    cout<<"Top Fretes:\n";
    for(const int& preco : top_fretes) cout << preco << " ";
    cout<<endl;
    
}

void Q4(){
    vector<string> rota = {"São Paulo", "Jundiaí", "Campinas", "Sorocaba"};
    vector<string> novasCidades = {"Itu", "Valinhos"};
    string identificarIndice = "Sorocaba";
    rota.insert(rota.end(), novasCidades.begin(), novasCidades.end());
    int pos = 0;
    bool flag = true;
    for(const string& cidade : rota){
        cout << cidade;
        if(cidade == identificarIndice) flag = false;
        if(flag) pos++;
        if(&rota.back() == &cidade) cout << "." << endl;
        else cout << ", ";
    } 
    cout<<"Indice de "<<identificarIndice<<": "<<pos<<endl;
    cout<<"Sorocaba é a "<<pos+1<<"a cidade da rota.\n";
    
}

void Q5(){
    vector<float> precos = {100.0, 200.0, 300.0, 500.0};
    vector<string> vinhos = {"Branco", "Tinto", "Rose", "Champagne"};
    cout<<"Lista de Produtos:\n";
    for(int i = 0; i < vinhos.size(); i++){
        cout << vinhos[i] << ": S" << fixed << setprecision(2) << precos[i] << ((i+1 == vinhos.size()) ? ".\n" : ";\n");
    }
    string nome_produto;
    while(true){
        cin >> nome_produto;
        if (find(vinhos.begin(), vinhos.end(), nome_produto) != vinhos.end()) {
                break;
        } else {
                cout << "O vinho " << nome_produto << " nao foi encontrado." << endl;
        }
        
    }
    cout << "Digite o novo preço do produto: ";
    float novo_preco = extractBillingValue();
    for(int i = 0; i < vinhos.size(); i++){
        if(nome_produto == vinhos[i]) precos[i] = novo_preco;
    }
    cout<<"Lista atualizada.\n";
    for(int i = 0; i < vinhos.size(); i++){
        cout << vinhos[i] << ": S" << fixed << setprecision(2) << precos[i] << ((i+1 == vinhos.size()) ? ".\n" : ",\n");
    }
}



int main(){
    //Q1();
    //Q2();
    //Q3();
    //Q4();
    Q5();
    return 0;
}